src/Hosting/TestHost/test/ClientHandlerTests.cs

```diff
@@ -103,6 +104,36 @@ public Task UserAgentHeaderWorks()
             return httpClient.GetAsync("http://example.com");
         }
 
+        [Fact]
+        public Task ContentLengthWithBodyWorks()
+        {
+            var contentBytes = Encoding.UTF8.GetBytes("This is a content!");
+            var handler = new ClientHandler(new PathString(""), new DummyApplication(context =>
+            {
+                Assert.Equal(contentBytes.LongLength, context.Request.ContentLength);
+
+                return Task.CompletedTask;
+            }));
+            var httpClient = new HttpClient(handler);
+            var content = new ByteArrayContent(contentBytes);
+
+            return httpClient.PostAsync("http://example.com", content);
+        }
+
+        [Fact]
+        public Task ContentLengthWithNoBodyWorks()
+        {
+            var handler = new ClientHandler(new PathString(""), new DummyApplication(context =>
+            {
+                Assert.Equal(0, context.Request.ContentLength);
```

Comment:
Interesting. Because TestServer defaults to a StreamContent(Stream.Null), and Stream.Null's default length is 0, all requests that don't specify a body will get `Content-Length: 0`.
https://github.com/dotnet/aspnetcore/blob/5980ad813e0044421b13ea70431452505262e946/src/Hosting/TestHost/src/ClientHandler.cs#L69

Reply:
Can you add a test where the client intentionally sets Transfer-Encoding: chunked? We don't want to automatically add a content-length in that scenario.

Reply:
Setting `Transfer-Encoding: chunked` for `ByteArrayContent` didn't do anything, the ContentLength was still 0.
But creating a content which return `false` on `TryComputeLength` would make `ContentLength` null. Is this good enough?